### PR TITLE
OI-74 Automatically follow group ideas.

### PR DESCRIPTION
**Pull request by nvelychenko**

### CODE REVIEW

File: config/install/rules.reaction.set_the_user_to_follow_the_idea_after_joining_the_group_.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+langcode: en
+status: true
+dependencies: {  }
+id: set_the_user_to_follow_the_idea_after_joining_the_group_
```
Comment (db-Matroskeen) [inline config/install/rules.reaction.set_the_user_to_follow_the_idea_after_joining_the_group_.yml]: Let's use better machine names.
For example: `follow_idea_after_joining_group`.

File: modules/openideal_user/openideal_user.module [hunk 2]
```diff
@@ -71,15 +76,15 @@ function openideal_user_entity_type_alter(array &$entity_types) {
  * Implements hook_preprocess_HOOK().
  */
 function openideal_user_preprocess_page(&$variables) {
-  if (\Drupal::currentUser()->isAuthenticated()) {
-    $user_id = \Drupal::currentUser()->id();
-    $user = \Drupal::entityTypeManager()->getStorage('user')->load($user_id);
-    $current_route = \Drupal::service('current_route_match')->getRouteName();
+  if (Drupal::currentUser()->isAuthenticated()) {
```
Comment (db-Matroskeen) [inline modules/openideal_user/openideal_user.module]: I'm curious, what is the purpose of changing `\Drupal` to `Drupal`?
In Drupal core, there is a standard way to use` \Drupal` although it's possible to use just Drupal in the procedural code: https://www.drupal.org/project/drupal/issues/2053489

So why are we not following it?

File: modules/openideal_user/openideal_user.module [hunk 3]
```diff
@@ -90,14 +95,43 @@ function openideal_user_preprocess_page(&$variables) {
...
+  }
+}
+
+/**
+ * Implements hook_flag_action_access().
+ *
+ * If user is Idea creator then restrict "unfollow" access.
+ */
+function openideal_idea_flag_action_access($action, FlagInterface $flag, AccountInterface $account, EntityInterface $flaggable = NULL) {
+  if ($flaggable && $flaggable instanceof NodeInterface && $action == 'unflag' && $flaggable->bundle() == 'idea') {
```
Comment (db-Matroskeen) [inline modules/openideal_user/openideal_user.module]: Let's restrict only for the original author of the idea (not co-authors).
It'll be enough to check if current user id is equal to a flagged entity.

File: modules/openideal_user/src/Plugin/RulesAction/FlagAction.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\flag\FlagServiceInterface;
+use Drupal\rules\Core\RulesActionBase;
+use Drupal\user\UserInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a 'Save entity' action.
```
Comment (db-Matroskeen) [inline modules/openideal_user/src/Plugin/RulesAction/FlagAction.php]: Good stuff. Can we publish this to flag module issues queue to add `Rules` support to the module?

**Review by db-Matroskeen — CHANGES_REQUESTED** [review]

File: modules/openideal_user/openideal_user.module (hunk 3 above)
Comment (nvelychenko) [reply modules/openideal_user/openideal_user.module]: But in the membership page, you can grant to the user "Author" role, what if there will be two authors or we should remove the ability to grant this role for more than one user?

File: modules/openideal_user/src/Plugin/RulesAction/FlagAction.php (hunk 4 above)
Comment (nvelychenko) [reply modules/openideal_user/src/Plugin/RulesAction/FlagAction.php]: Reworked the module to be more "flexible" without hardcoded details. Please, take a look.

File: modules/openideal_user/openideal_user.module (hunk 2 above)
Comment (nvelychenko) [reply modules/openideal_user/openideal_user.module]: My bad, it's the "silent code cleanup" in PHPSTORM.

File: modules/openideal_user/src/Plugin/RulesAction/FlagAction.php (hunk 4 above)
Comment (db-Matroskeen) [reply modules/openideal_user/src/Plugin/RulesAction/FlagAction.php]: It looks better, but we'll have to adjust some details before publishing it for review.
We'll get back to this later.

File: modules/openideal_user/openideal_user.module [hunk 5]
```diff
@@ -90,14 +95,34 @@ function openideal_user_preprocess_page(&$variables) {
...
+ */
+function openideal_idea_flag_action_access($action, FlagInterface $flag, AccountInterface $account, EntityInterface $flaggable = NULL) {
+  if ($flaggable
+    && $flaggable instanceof NodeInterface
+    && $action == 'unflag'
+    && $flaggable->bundle() == 'idea'
+    && $flaggable->uid->target_id == $account->id()) {
+    return AccessResult::forbidden();
+  }
+  return AccessResult::allowed();
```
Comment (db-Matroskeen) [inline modules/openideal_user/openideal_user.module]: If there is no need to restrict or allow some specific access, we have to use `AccessResult::neutral()`.

**Review by db-Matroskeen — CHANGES_REQUESTED** [review]

**Pull request merged**